nemo/lightning/megatron_parallel.py

```diff
@@ -475,6 +482,37 @@ def infer_num_microbatches(self, data: Union[DataT, Iterator[DataT], List[Iterat
 
         raise ValueError("Cannot infer `num_microbatches` from data, please specify it manually")
 
+    def init_tensor_parallel_overlap(self, micro_batch_size, sequence_length):
+        import transformer_engine
+        from megatron.core import parallel_state
+
+        hidden_size = self.config.hidden_size
+        fp8 = self.config.fp8
+        ub_cfgs = self.config.tp_comm_overlap_cfg
+        cp_size = parallel_state.get_context_parallel_world_size()
+        tp_size = parallel_state.get_tensor_model_parallel_world_size()
+
+        input_shape = [
+            sequence_length * micro_batch_size // cp_size,
+            hidden_size,
+        ]
+
+        try:
+            if ub_cfgs is None:
+                logging.warning(
+                    "Tensor parallel overlap: No overlap config provided. Initializing TP comm overlap with the default config."
+                )
+            transformer_engine.pytorch.module.base.initialize_ub(
+                shape=input_shape,
+                tp_size=tp_size,
+                use_fp8=fp8,
+                ub_cfgs=ub_cfgs,
+            )
+        except Exception as error:
+            raise Exception(f"Tensor parallel overlap: userbuffer initialization failed with {error}")
+
+        self.tensor_parallel_overlap_need_init = False
```

Comment:
Why is this required ?

Reply:
For the init we need to microbatch size and sequence length, and in megatron_parallel.py it seems this is only available in [in the forward function](https://github.com/NVIDIA/NeMo/pull/10209/files/95267ec814f617adc00564b10fef622cbbdb8d6b#diff-3a143831f2dddc93960ceef806f3dd2b06b29de548b3ecc0669cf3ab6969e85eR238)

So I have this so that init only occurs once there, maybe there is somewhere init should go so that we woudn't need this?